src/bidiMapper/domains/context/logEntry.ts

```diff
@@ -0,0 +1,276 @@
+      throw Error('Invalid value type: ' + arg.toString());
+    }
+
+    if (arg.type === 'number' && typeof (arg.value, Number)) {
+      return arg.value.toString();
+    }
+
+    if (arg.type === 'string' && typeof (arg.value, String)) {
+      return '"' + arg.value.toString() + '"';
+    }
+
+    if (arg.type === 'object' && typeof (arg.value, Object)) {
+      return (
+        '{"' +
+        (arg.value as any[])[0][0] +
+        '": ' +
+        this._getSingleRemoteValueText((arg.value as any[])[0][1]) +
+        '}'
+      );
+    }
+
+    throw Error('Invalid value type: ' + arg.toString());
+  }
+
+  private _getRemoteValuesText(
+    args: CommonDataTypes.RemoteValue[],
+    formatText: boolean
+  ): string {
+    // if args[0] is a format specifier, format the args as output
+    if (
+      args[0].type === 'string' &&
+      this._isFormmatSpecifier(args[0].value.toString()) &&
+      formatText
+    ) {
+      return this._formatter(args);
+    }
+
+    // if args[0] is not a format specifier, just join the args with \u0020
+    return args
+      .map((arg) => (arg as { type: string; value: any }).value.toString())
+      .join('\u0020');
+  }
+
+  // convert CDP StackTrace object to Bidi StackTrace object
+  private _getBidiStackTrace(
+    cdpStackTrace: Protocol.Runtime.StackTrace | undefined
+  ): Script.StackTrace | undefined {
+    const stackFrames = cdpStackTrace?.callFrames.map((callFrame) => {
+      return {
+        columnNumber: callFrame.columnNumber,
+        functionName: callFrame.functionName,
+        lineNumber: callFrame.lineNumber,
+        url: callFrame.url,
+      };
+    });
+
+    const bidiStackTrace = stackFrames
+      ? {
+          callFrames: stackFrames,
+        }
+      : undefined;
+
+    return bidiStackTrace;
+  }
+
+  private _initializeLogEntryAddedEventListener() {
+    this._cdpClient.Runtime.on('consoleAPICalled', async (params) => {
+      const args = await Promise.all(
+        params.args.map(async (arg) => {
+          return await this._serializeCdpObject(arg);
+        })
+      );
+
+      await this._bidiServer.sendMessage({
+        method: 'log.entryAdded',
+        params: {
+          level: this._getLogLevel(params.type),
+          text: this._getRemoteValuesText(args, true),
+          timestamp: params.timestamp,
+          stackTrace: this._getBidiStackTrace(params.stackTrace),
+          type: 'console',
+          method: params.type,
+          realm: this._contextId,
+          args: args,
+        },
+      });
+    });
+  }
+
+  // TODO sadym: `dummyContextObject` needed for the running context.
+  // Use the proper `executionContextId` instead:
+  // https://github.com/GoogleChromeLabs/chromium-bidi/issues/52
+  private async _getDummyContextId(): Promise<string> {
+    const dummyContextObject = await this._cdpClient.Runtime.evaluate({
+      expression: '(()=>{return {}})()',
+    });
+    return dummyContextObject.result.objectId!;
+  }
```

Comment:
making the same trick again on the sabe context sounds an overkill. Can we keep the serialization in one place?

Reply:
moved the function getDummyContextId to contextHelper.ts